### PR TITLE
`PercentageScore` heuristic for significant_terms

### LINKED ISSUE #9720: significant_terms should provide a simple percentage scoring heuristic
As discussed with @markharwood on the [mailing list](https://groups.google.com/d/msg/elasticsearch/1ojltqSRdhA/7tvVKKNTWe4J), it would be nice if `significant_terms` provided a scoring heuristic that is simply `doc_count` divided by `bg_count`.

In some contexts, scoring needs to be easy to explain to users, and percentages are something most of us can understand.

### PR DESCRIPTION
**Pull request by markharwood**
Provides simple “per-capita” type scoring measure for use on higher-frequency terms

Closes #9720

### CODE REVIEW

**Comment by markharwood** [comment]
@brwe This seemed a simple but potentially common heuristic so I added a native implementation rather than relying on a scripted solution. I added a cautionary note in the reference docs about use on fields with low-frequency terms after some trials on real data.

File: src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/SignificanceHeuristic.java [hunk 1]
```diff
@@ -42,10 +42,10 @@ protected void checkFrequencyValidity(long subsetFreq, long subsetSize, long sup
             throw new ElasticsearchIllegalArgumentException("Frequencies of subset and superset must be positive in " + scoreFunctionName + ".getScore()");
         }
         if (subsetFreq > subsetSize) {
-            throw new ElasticsearchIllegalArgumentException("subsetFreq > subsetSize, in JLHScore.score(..)");
+            throw new ElasticsearchIllegalArgumentException("subsetFreq > subsetSize, in " + scoreFunctionName + ".score(..)");
```
Comment (colings86) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/SignificanceHeuristic.java]: Do we need the `.score(..)` here? For the user all they need to know is the score function that failed. The method is probably not relevant to the user in the message and will be in the stack trace for a dev investigating anyway

File: src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.QueryParsingException;
+
+import java.io.IOException;
+
+public class PercentageScore extends SignificanceHeuristic {
+
+    public static final PercentageScore INSTANCE = new PercentageScore();
+
+    protected static final String[] NAMES = {"percentage"};
```
Comment (colings86) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java]: Does this need to be a String[]? e seem to only use it in one place where we just get the first element anyway so would it not be better as a plain String?

Comment (markharwood) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java]: The SignificanceHeuristicParser base class has a getNames() method that requires an array of names (presumably to allow for alternatives) so this is returned there

File: src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * Indicates the significance of a term in a sample by determining what percentage
+     * of all occurrences of a term are found in the sample. 
+     */
+    @Override
+    public double getScore(long subsetFreq, long subsetSize, long supersetFreq, long supersetSize) {
+        checkFrequencyValidity(subsetFreq, subsetSize, supersetFreq, supersetSize, "PercentageScore");
+        if (supersetFreq == 0) {
```
Comment (brwe) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java]: If subsetFreq is > 0 and supersetFreq = 0, should the user not be warned?

File: src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java (hunk 2 above)
Comment (colings86) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java]: Ah ok, I missing that method below, sorry.

File: src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java (hunk 3 above)
Comment (markharwood) [inline src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/PercentageScore.java]: I've followed the logic we've used elsewhere. I think this allows for a case where a background_filter is not a strict superset e.g. the foreground is you search for all tweets with "JeSuisCharlie" and set a background of "language:fr" to tune out everyday French like "vous" and hopefully focus in on the Charlie Hebdo related keywords. The background stats are used in that example as a rough backdrop and we choose not to error if the foreground has a term not in the background.

**Comment by jpountz** [comment]
LGTM

**Comment by clintongormley** [comment]
Merged in https://github.com/elastic/elasticsearch/commit/4d2649261211486b192d23d0d0e9415b4dffa7ea

**Pull request closed without merging**